drivers/usb/hid-keyboard.lisp

```diff
@@ -1,19 +1,377 @@
+    :Currency-Unit :Currency-Sub-Unit
+    :KP-Left-Paren :KP-Right-Paren :KP-Left-Brace :KP-Right-Brace
+    :KP-Tab :KP-Backspace
+    :KP-A :KP-B :KP-C :KP-D :KP-E :KP-F
+    :KP-XOR :KP-^ :KP-% :KP-< :KP-> :KP-& :KP-&& :KP-Bar :KP-Bar-Bar
+    :KP-Colon :KP-Pound :KP-Space :KP-@ :KP-!
+    :KP-Mem-Store :KP-Mem-Recall :KP-Mem-Clear
+    :KP-Mem-Add :KP-Mem-Minus :KP-Mem-Multiply :KP-Mem-Divide
+    :KP-Plus-Minus :KP-Clear :KP-Clear-Entry
+    :KP-Binary :KP-Octal :KP-Decimal :KP-Hexadecimal
+    :reserved-DE :reserved-DF
+    :Left-Control :Left-Shift :Left-Alt :Left-GUI
+    :Right-Control :Right-Shift :Right-Alt :Right-GUI
+    :reserved-E8 :reserved-E9 :reserved-EA :reserved-EB
+    :reserved-EC :reserved-ED :reserved-EE :reserved-EF
+    :reserved-F0 :reserved-F1 :reserved-F2 :reserved-F3
+    :reserved-F4 :reserved-F5 :reserved-F6 :reserved-F7
+    :reserved-F8 :reserved-F9 :reserved-FA :reserved-FB
+    :reserved-FC :reserved-FD :reserved-FE :reserved-FF))
 
+;; submit key function name, the compositor package may not defined
+;; when USB drivers are loaded, the symbol is generated when a hid
+;; keyboard is probed.
+(defvar *submit-key* nil)
 
-(defun probe-hid-keyboard (usbd device configs state)
-  (let ((endpoint (make-hid-endpt :type :keyboard)))
+;;======================================================================
+;; HID Keyboard interrupt handler
+;;======================================================================
+(defvar *keyboard-ints* NIL) ;; for debug
+
+(defstruct (hid-keyboard-endpt (:include hid-endpt) (:conc-name hid-endpt-))
+  pressed-keys
+  pressed-modifiers)
+
+(defun keyboard-int-callback (driver endpoint-num status length buf)
+  (unwind-protect
+       (cond ((eq status :success)
+              (let ((endpoint (aref (hid-driver-endpoints driver) endpoint-num)))
+                (funcall (hid-endpt-function endpoint) endpoint length buf)))
+             (T
+              (format sys.int::*cold-stream*
+                      "Interrupt error ~A on endpoint number ~D~%"
+                      status endpoint-num)
+              (with-trace-level (1)
+                (format sys.int::*cold-stream*
+                        "length: ~D~%buf: ~S~%" length buf))))
+    (with-trace-level (7)
+      (push (format nil "~D: ~A" length buf) *keyboard-ints*))
+    (free-buffer buf)))
+
+;;======================================================================
+;; Code which generates function to process interrupt buffer data
+;;======================================================================
+(defvar *modifier-buttons*
+  ;; Modifier buttons can either be :button-e[0-7] or :<keyword>
+  ;; depending on how the HID descriptor is defined, so we have to
+  ;; check for both cases.
+  '(:button-e0 :button-e1 :button-e2 :button-e3
+    :button-e4 :button-e5 :button-e6 :button-e7
+    :left-control :left-shift :left-alt :left-gui
+    :right-control :right-shift :right-alt :right-gui))
+
+(defvar *modifier->char*
```

Comment:
How does this work with layouts that have more modifiers?  I am using neo2 layout, which has Shift, Mod3, Mod4, and I use Control, Meta, Super, Hyper as command modifiers.  xkb in X can support this.  This is also relevant if you want to get a complete Space Cadet keyboard as in the Lisp Machines.

Reply:
Sorry, it doesn't handle more modifiers. I'm not sure the HID class driver can support more modifiers - there doesn't seem to be button codes defined for additional modifiers (USB HID Usage Tables doc). I would like the driver to be as general as possible, but your keyboard may need a device specific driver. (The mezzano USB driver does support device specific drivers, but we don't have any yet.)

To update the code, I need to see the report descriptor(s) that your keyboard provides. You can get that on mezzano, but it requires you to have a USB-2 controller (I'm still working on the USB-3 controller driver) and merge in the PR, turn on debug tracing that I've included in the code and then connect your keyboard.

I can give you more details on how to turn on logging if you want to pursue this.